### PR TITLE
Implemented `urlFilter` feature.

### PR DESCRIPTION
**Pull request by marklagendijk**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -118,3 +119,19 @@ scraper.scrape({
   maxDepth: 1
 }).then(console.log).catch(console.log);
 ```
+
+#### Example 3. Recursive downloading of entire website
```
Comment (s0ph1e) [inline README.md]: It is hard to see difference between Example 3 and Example 2

I suggest to make this example only for `urlFilter` (without `recursive` and `maxDepth`) to highlight this feature 
For example, title may be like **Filtering out external resources**

File: lib/scraper.js [hunk 2]
```diff
@@ -99,6 +99,10 @@ Scraper.prototype.getResourceHandler = function getHandler (resource) {
 Scraper.prototype.loadResource = function loadResource (resource) {
 	var self = this;
 
+	if(!self.options.urlFilter(resource.url)){
+		return Promise.resolve(null);
```
Comment (s0ph1e) [inline lib/scraper.js]: I'm not sure but I think we can return rejected promise so modifications in file-handlers will be not needed

Comment (marklagendijk) [inline lib/scraper.js]: I don't think that would be right. Rejecting a promise is the same as throwing an exception, and should only be used for exception flows.
Also, if you would do it, you would still need to add `.catch` in the file-handlers.

Btw, did you know that the following are the same?
``` javascript
return doSomethingWhichReturnsAPromise()
  .then(function(item){
    return Promise.resolve(item.property);
  });

return doSomethingWhichReturnsAPromise()
  .then(function(item){
    return item.property;
  });
// Inside a then you should just return a value, if the logic is synchronous. 
// You only need to return a promise if it is asynchronous.
```
And this:
``` javascript
return doSomethingWhichReturnsAPromise()
  .then(function(item){
    return Promise.reject(new Error('Something is wrong!'));
  });

return doSomethingWhichReturnsAPromise()
  .then(function(item){
    throw new Error('Something is wrong!');
  });
// Inside a then you should just throw an error, if the logic is synchronous. 
// You only need to return a rejected promise if it is asynchronous.
```

Comment (marklagendijk) [inline lib/scraper.js]: The problem with the current code is that the following things are too tightly coupled:

1. Reading resource urls from the code (HTML, CSS).
2. Determining whether the urls should be processed (`urlFilter`).
3. Determining the location where the file should be saved.
4. Retrieving and saving the resources.
5. Updating the urls based on the location in which they where saved.

To properly solve this, the code would need huge refactoring. You would want to end up with something like this (using ES6 syntax for brevity):
``` javascript
resource
  .createChildResources() // reads the child resource urls using the right handler, based on the resource type, and constructs resources from them.
  .then(childResources => _.filter(childResources, childResource => context.options.urlFilter(childResource.url)));
  .then(childResources => _.forEach(childResource => childResource.updateUrlInParent()))
  .then(childResources, Promise.all(_.map(childResources, childResource => loadAndSaveResource(childResource)))
```

Comment (s0ph1e) [inline lib/scraper.js]: Yep, I agree it needs huge refactoring
Code was written long time ago without any architecture as quick script to solve my needs and then migrated to module
But it works and people use it. I'm afraid of doing huge refactor now

**Comment by s0ph1e** [comment]
Looks good 
Thanks @marklagendijk for your contribution
I want to test it locally tonight and after it I'll merge PR

File: test/unit/file-handlers/html-test.js [hunk 3]
```diff
@@ -322,6 +359,40 @@ describe('Html handler', function () {
 			}).catch(done);
 		});
 
+		it('should not replace the replace the sources of img tags with srcset attribute, for which loadResource returned null', function (done) {
```
Comment (s0ph1e) [inline test/unit/file-handlers/html-test.js]: Typo: redundant **the replace**

Should be `should not replace the sources of img tags with srcset attribute, for which loadResource returned null`

File: test/unit/scraper-test.js [hunk 4]
```diff
@@ -578,6 +578,50 @@ describe('Scraper', function () {
...
+					lr.getText().should.be.eql('OK');
+
+					var text = fs.readFileSync(path.join(testDirname, lr.getFilename())).toString();
+					text.should.be.eql(lr.getText());
+					done();
+				});
+			}).catch(done);
+		});
+
+		it('should not return an promise resolved with null if the urlFilter returns false', function(done){
```
Comment (s0ph1e) [inline test/unit/scraper-test.js]: Typo: seems **not** is not needed here

File: test/unit/scraper-test.js [hunk 5]
```diff
@@ -578,6 +578,50 @@ describe('Scraper', function () {
...
+			var s = new Scraper({
+				urls: ['http://example.com', 'http://google.com'],
```
Comment (s0ph1e) [inline test/unit/scraper-test.js]: Why do we need 2 urls here?

**Comment by s0ph1e** [comment]
:+1: 
will be merged after minor typos fixes

**Pull request merged**